clang/include/clang/CAS/IncludeTree.h

```diff
@@ -0,0 +1,264 @@
+    }
+    return isValid(*Node);
+  }
+
+private:
+  friend class IncludeTreeBase;
+  friend class IncludeTree;
+  friend class IncludeTreeRoot;
+
+  explicit IncludeFile(ObjectProxy Node) : IncludeTreeBase(std::move(Node)) {
+    assert(isValid(*this));
+  }
+};
+
+/// Represents a DAG of included files by the preprocessor.
+/// Each node in the DAG represents a particular inclusion of a file that
+/// encompasses inclusions of other files as sub-trees, along with all the
+/// \p __has_include() preprocessor checks that occurred during preprocessing
+/// of that file.
+class IncludeTree : public IncludeTreeBase<IncludeTree> {
+public:
+  static constexpr StringRef getNodeKind() { return "Tree"; }
+
+  Expected<IncludeFile> getBaseFile() {
+    auto Node = getCAS().getProxy(getReference(0));
+    if (!Node)
+      return Node.takeError();
+    return IncludeFile(std::move(*Node));
+  }
+
+  /// The include file that resulted in this include-tree.
+  ObjectRef getBaseFileRef() const { return getReference(0); }
+
+  Expected<IncludeFile::FileInfo> getBaseFileInfo() {
+    auto File = getBaseFile();
+    if (!File)
+      return File.takeError();
+    return File->getFileInfo();
+  }
+
+  SrcMgr::CharacteristicKind getFileCharacteristic() const {
+    return (SrcMgr::CharacteristicKind)dataSkippingIncludes().front();
+  }
+
+  size_t getNumIncludes() const { return getNumReferences() - 1; }
+
+  ObjectRef getIncludeRef(size_t I) const {
+    assert(I < getNumIncludes());
+    return getReference(I + 1);
+  }
+
+  /// The sub-include-trees of included files, in the order that they occurred.
+  Expected<IncludeTree> getInclude(size_t I) {
+    return getInclude(getIncludeRef(I));
+  }
+
+  /// The source byte offset for a particular include, pointing to the beginning
+  /// of the line just after the #include directive. The offset represents the
+  /// location at which point the include-tree would have been processed by the
+  /// preprocessor and parser.
+  ///
+  /// For example:
+  /// \code
+  ///   #include "a.h" -> include-tree("a.h")
+  ///   | <- include-tree-offset("a.h")
+  /// \endcode
+  ///
+  /// Using the "after #include" offset makes it trivial to identify the part
+  /// of the source file that encompasses a set of include-trees (for example in
+  /// the case where want to identify the "includes preamble" of the main file.
+  unsigned getIncludeOffset(size_t I) const;
+
+  /// The \p __has_include() preprocessor checks, in the order that they
+  /// occurred. The source offsets for the checks are not tracked, "replaying"
+  /// the include-tree depends on the invariant that the same exact checks will
+  /// occur in the same order.
+  bool getCheckResult(size_t I) const;
+
+  llvm::Error forEachInclude(
+      llvm::function_ref<llvm::Error(std::pair<IncludeTree, unsigned>)>
+          Callback);
+
+  static Expected<IncludeTree>
+  create(CASDB &DB, SrcMgr::CharacteristicKind FileCharacteristic,
+         ObjectRef IncludedBy,
```

Comment:
Maybe rename this parameter to `BaseFile` to match `getBaseFile` etc?  It kind of sounds like it is the parent node in the IncludeTree.

Reply:
Done 👍